### PR TITLE
PolarFly and PolarStar topologies

### PR DESCRIPTION
**Pull request by kartiklakhotia**
PolarFly and PolarStar are low-diameter network topologies with state-of-the-art scalability.  We have added the following for these topologies (in Merlin component):

1. Generators
2. Routing functions
3. Test files with golden (reference) output

References:
1. Lakhotia, Kartik, Maciej Besta, Laura Monroe, Kelly Isham, Patrick Iff, Torsten Hoefler, and Fabrizio Petrini. "PolarFly: a cost-effective and flexible low-diameter topology." In Proceedings of the International Conference on High Performance Computing, Networking, Storage and Analysis, 2022.

2. Lakhotia, Kartik, Laura Monroe, Kelly Isham, Maciej Besta, Nils Blach, Torsten Hoefler, and Fabrizio Petrini. "PolarStar: Expanding the Scalability Horizon of Diameter-3 Networks." arXiv preprint arXiv:2302.07217 (2023).

### CODE REVIEW

File: src/sst/elements/merlin/Makefile.am [hunk 1]
```diff
@@ -50,6 +50,11 @@ libmerlin_la_SOURCES = \
 	topology/singlerouter.cc \
 	topology/hyperx.h \
 	topology/hyperx.cc \
+	topology/polarfly.cc \
+	topology/polarfly.h \
+	topology/polarstar.cc \
+	topology/polarstar.h \
+	topology/polar.H \
```
Comment (feldergast) [inline src/sst/elements/merlin/Makefile.am:57]: I don't see a polar.H file in the changed files.  Either this is here on accident or you're missing a file.  Also, if you're missing the file, please make sure it is polar.h and not polar.H.

File: src/sst/elements/merlin/topology/polarfly.h [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        topo_polarfly,
+        "merlin",
+        "polarfly",
+        SST_ELI_ELEMENT_VERSION(1,0,0),
+        "PolarFly: A Cost-effective Two-diameter highly scalable topology object",
+        SST::Merlin::Topology)
+
+    SST_ELI_DOCUMENT_PARAMS(
+        // Parameters needed for use with old merlin python module
+        {"polarfly.q", "size of the Gallois field. parameter should be a prime power."},
```
Comment (feldergast) [inline src/sst/elements/merlin/topology/polarfly.h:136]: The polarfly.* params are not needed.  I'm sure you copied this from existing topologies, but these are in those topologies for backward compatibility to an older python module implementation.  They won't be needed for this one.  All of these should be removed.

File: src/sst/elements/merlin/topology/polarstar.h [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+        topo_polarstar,
+        "merlin",
+        "polarstar",
+        SST_ELI_ELEMENT_VERSION(1,0,0),
+        "PolarStar: Highly Scalable Diameter-3 Topology Object",
+        SST::Merlin::Topology)
+
+    SST_ELI_DOCUMENT_PARAMS(
+        // Parameters needed for use with old merlin python module
+        {"polarstar.d", "Network Radix per router"},
```
Comment (feldergast) [inline src/sst/elements/merlin/topology/polarstar.h:134]: The polarstar.* params are not needed.  I'm sure you copied this from existing topologies, but these are in those topologies for backward compatibility to an older python module implementation.  They won't be needed for this one.

**Review by feldergast — CHANGES_REQUESTED** [review]
In addition to addressing the comments inlined in the files changed listing, please rebase against the latest devel branch of the element repo.  We had a recent bug slip through testing that was recently fixed, and you will need to make sure you have that fix or you won't pass autotesting.

File: src/sst/elements/merlin/topology/polarstar.h (hunk 3 above)
Comment (kartiklakhotia) [reply src/sst/elements/merlin/topology/polarstar.h:134]: > In addition to addressing the comments inlined in the files changed listing, please rebase against the latest devel branch of the element repo. We had a recent bug slip through testing that was recently fixed, and you will need to make sure you have that fix or you won't pass autotesting.

Should we create a pull request for the devel branch or the master?

Comment (jpkenny) [reply src/sst/elements/merlin/topology/polarstar.h:134]: Please create the pull request to devel.
Documentation here may be useful: 
https://github.com/sstsimulator/sst-elements/blob/master/CONTRIBUTING.md
Thanks

Comment (kartiklakhotia) [reply src/sst/elements/merlin/topology/polarstar.h]: We are getting an error with SST_ELI_REGISTER_SUBCOMPONENT while compiling Merlin in the devel branch:

In file included from merlin.cc:25:
hr_router/xbar_arb_rr.h: At global scope:
hr_router/xbar_arb_rr.h:40:9: error: expected identifier before string constant
   40 |         "merlin",
      |         ^~~~~~~~
hr_router/xbar_arb_rr.h:40:9: error: expected ‘,’ or ‘...’ before string constant
hr_router/xbar_arb_rr.h:38:5: error: ISO C++ forbids declaration of ‘SST_ELI_REGISTER_SUBCOMPONENT’ with no type [-fpermissive]
   38 |     SST_ELI_REGISTER_SUBCOMPONENT(
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
hr_router/xbar_arb_rr.h:45:5: error: expected ‘;’ at end of member declaration
   45 |     )
      |     ^
      |      ;

**Pull request closed without merging**